src/alchemlyb/parsing/namd.py

```
@@ -25,72 +82,136 @@ def extract_u_nk(fep_file, T):
+                # within the same file, then complain. This can happen if truncated fepout files
+                # are presented in the wrong order.
+                if l[0] == '#NEW':
+                    lambda1_at_start, lambda2_at_start = float(l[6]), float(l[8])
+                    lambda_idws_at_start = float(l[10]) if 'LAMBDA_IDWS' in l else None
+                    has_idws = True if lambda_idws_at_start is not None else False
+
+                # this line marks end of window; dump data into dataframe
+                if l[0] == '#Free':
+                    # extract lambda values for finished window
+                    # lambda1 = sampling lambda (row), lambda2 = comparison lambda (col)
+                    lambda1 = float(l[7])
+                    lambda2 = float(l[8])
+                    lambda1_idx = all_lambdas.index(lambda1)
+                    if has_idws is True and lambda1_idx > 0:
+                        lambda_idws = all_lambdas[lambda1_idx - 1]
+                    else:
+                        lambda_idws = None
+
+                    # If the lambdas are not what we thought they would be, return None, ensuring the calculation
+                    # fails. This can happen if fepouts where one window spans multiple fepouts are processed out of order
+                    if lambda1_at_start is not None \
+                        and (lambda1, lambda2, lambda_idws) != (lambda1_at_start, lambda2_at_start, lambda_idws_at_start):
+                        print("namd.py: extract_u_nk: Error: Lambdas changed unexpectedly while processing", fep_file)
+                        print(f"namd.py: extract_u_nk: Error: l1, l2, lidws: {lambda1_at_start}, {lambda2_at_start}, {lambda_idws_at_start} changed to {lambda1}, {lambda2}, {lambda_idws}")
+                        print(f"namd.py: extract_u_nk: Error: fep_file = {fep_file}; has_idws = {has_idws}")
```

Do not use `print()` in library code. Use a logger for *alchemlyb.parsers.NAMD* (see what the Amber parser does https://github.com/alchemistry/alchemlyb/blob/b068776354055115da6d52af24664ba3938985cc/src/alchemlyb/parsing/amber.py#L21 ).

Instead of returning `None`, raise the exception here; `ValueError` could be appropriate unless you have another idea.